### PR TITLE
Handle stateful properties in xfem moving interface

### LINKED ISSUE #15437: Add AD capability for existing BiMaterial XFEM classes
## Reason
Moving interface problems using XFEM and level set capabilities cannot currently be used with material classes using AD.

## Design
Duplicate the BiMaterial classes with AD counterparts

## Impact
Extends the use of XFEM-level set simulations to material models coded only in AD

ping @jiangwen84

### LINKED ISSUE #15817: xfem moving interface + stateful properties
## Reason

Currently XFEM moving interface is not setup to handle stateful properties.

## Design

Store the stateful properties before healing, and set them back after re-cut.

## Impact

No change on existing APIs, but will enable us to do moving interface with finite deformation etc.

### PR DESCRIPTION
**Pull request by hugary1995**
This is a reopening of stale PR #15818.
This PR handles stateful properties when an XFEM cut is healed and re-cut.

closes #15817

### CODE REVIEW

File: modules/xfem/include/base/XFEM.h [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include "libmesh/quadrature.h"
 
 #include "GeometricCutUserObject.h"
+#include "LevelSetCutUserObject.h"
```
Comment (jiangwen84) [inline modules/xfem/include/base/XFEM.h:21]: Do we have to put `level set` in XFEM? I understand that you need `level set` to tell you the material belongs to which side. Can we make it not be too specific to `level set`? i.e. `mesh cut uo`.

**Comment by hugary1995** [comment]
Pinging @bwspenc .

File: modules/xfem/include/base/XFEM.h (hunk 1 above)
Comment (hugary1995) [reply modules/xfem/include/base/XFEM.h:21]: Yeah I'll talk to you to see how we should design this.

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C [hunk 2]
```diff
@@ -25,8 +25,8 @@ ComputeFiniteStrainElasticStress::ComputeFiniteStrainElasticStress(
     GuaranteeConsumer(this),
     _elasticity_tensor_name(_base_name + "elasticity_tensor"),
     _elasticity_tensor(getMaterialPropertyByName<RankFourTensor>(_elasticity_tensor_name)),
-    _rotation_total(declareProperty<RankTwoTensor>("rotation_total")),
-    _rotation_total_old(getMaterialPropertyOldByName<RankTwoTensor>("rotation_total")),
+    _rotation_total(declareProperty<RankTwoTensor>(_base_name + "rotation_total")),
+    _rotation_total_old(getMaterialPropertyOldByName<RankTwoTensor>(_base_name + "rotation_total")),
```
Comment (recuero) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C:29]: Necessary addition of `_base_name`

Comment (recuero) [reply modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C:29]: Could you also add it to `ADComputeFinitStrainElasticStress`?

Comment (hugary1995) [reply modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C:29]: Oh I forgot that regular and AD are still in separate files here. I just did. Thanks for catching that.

File: modules/xfem/test/tests/moving_interface/moving_ad_bimaterial_finite_strain.i [hunk 3]
```diff
@@ -0,0 +1,271 @@
+# This test is for two layer materials with different youngs modulus with AD
+# The global stress is determined by switching the stress based on level set values
+# The material interface is marked by a level set function
+# The two layer materials are glued together
+# This case is also meant to test for a bug in moving interfaces on displaced meshes
```
Comment (bwspenc) [inline modules/xfem/test/tests/moving_interface/moving_ad_bimaterial_finite_strain.i:5]: Since I assume the bug doesn't exist any more, I would reword this to not mention the bug, but say what the scenario that it tests to make sure works is.

File: modules/xfem/test/tests/moving_interface/moving_bimaterial_finite_strain.i [hunk 4]
```diff
@@ -0,0 +1,308 @@
+# This test is for two layer materials with different youngs modulus
+# The global stress is determined by switching the stress based on level set values
+# The material interface is marked by a level set function
+# The two layer materials are glued together
+# This case is also meant to test for a bug in moving interfaces on displaced meshes
+# It should fail during the healing step of the 2nd timestep if the bug is present.
```
Comment (bwspenc) [inline modules/xfem/test/tests/moving_interface/moving_bimaterial_finite_strain.i:6]: Same thing here about the bug.

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/xfem/src/base/XFEM.C [hunk 5]
```diff
@@ -1369,6 +1361,104 @@ XFEM::cutMeshWithEFA(NonlinearSystemBase & nl, AuxiliarySystem & aux)
         _displaced_mesh->boundary_info->add_edge(libmesh_elem2, edge, parent_boundary_ids);
       }
     }
+
+    // TODO: The 0 here is the thread ID.  Need to sort out how to do this correctly
```
[the remark this answers is not in the document]
Comment (hugary1995) [reply modules/xfem/src/base/XFEM.C:1365]: I am not sure. This comment was there before I believe.

File: modules/xfem/include/base/XFEM.h (hunk 1 above)
Comment (hugary1995) [reply modules/xfem/include/base/XFEM.h]: Now this should be general enough to handle all geometric cuts. The related method is called `getConsistentSideBoolean` in the `GeometricCutUserObject` base class, which should be overridden by child classes. If it is not overridden AND healing is turned on AND there are stateful properties, it will throw an error.

**Comment by hugary1995** [comment]
Can someone help me invalidate the failing tests due to conda?

**Comment by loganharbour** [comment]
> Can someone help me invalidate the failing tests due to conda?

Looks good now.

File: modules/xfem/include/userobjects/GeometricCutUserObject.h [hunk 6]
```diff
@@ -182,6 +182,18 @@ class GeometricCutUserObject : public CrackFrontPointsProvider
    */
   bool shouldHealMesh() const { return _heal_always; };
 
+  /**
+   * Get a consistent boolean telling which side the node belongs to relative to the cut
```
Comment (bwspenc) [inline modules/xfem/include/userobjects/GeometricCutUserObject.h:186]: What do you mean by a 'consistent boolean'? I would suggest maybe renaming this class (or explaining what the consistent boolean is).

File: modules/xfem/include/userobjects/LevelSetCutUserObject.h [hunk 7]
```diff
@@ -38,6 +38,8 @@ class LevelSetCutUserObject : public GeometricCutUserObject
   virtual const std::vector<Point>
   getCrackFrontPoints(unsigned int num_crack_front_points) const override;
 
+  virtual bool getConsistentSideBoolean(const Node * node) const override;
```
Comment (bwspenc) [inline modules/xfem/include/userobjects/LevelSetCutUserObject.h:41]: Add doxygen

File: modules/xfem/src/base/XFEM.C (hunk 5 above)
[the remark this answers is not in the document]
Comment (bwspenc) [reply modules/xfem/src/base/XFEM.C:1363]: I'd delete that comment then.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Really close!  Just a few little things to fix.

File: modules/xfem/include/userobjects/LevelSetCutUserObject.h (hunk 7 above)
Comment (hugary1995) [reply modules/xfem/include/userobjects/LevelSetCutUserObject.h:41]: This is overriding the function in the parent class. Do I still need to add doxygen here?

File: modules/xfem/include/userobjects/GeometricCutUserObject.h (hunk 6 above)
Comment (hugary1995) [reply modules/xfem/include/userobjects/GeometricCutUserObject.h:186]: I want the function to return a boolean that is persistent across the entire simulation. The boolean itself should not contain any realistic information though. Can you teach me the English here, should it "consistent" or "persistent"?

File: modules/xfem/include/userobjects/LevelSetCutUserObject.h (hunk 7 above)
Comment (hugary1995) [reply modules/xfem/include/userobjects/LevelSetCutUserObject.h]: Well I added a doxygen anyways.

File: modules/xfem/src/base/XFEM.C [hunk 8]
```diff
@@ -2042,3 +2125,84 @@ XFEM::getNodeSolutionDofs(const Node * node, SystemBase & sys) const
...
+    for (auto bdid : parent_boundary_ids)
+    {
+      if (_fe_problem->needBoundaryMaterialOnSide(bdid, 0))
+        (*_bnd_material_data)[0]->copy(*cut_elem, props, props_old, props_older, side);
+    }
+  }
+}
+
+bool
+XFEM::getElemConsistentSideBoolean(const Elem * parent_elem,
```
Comment (bwspenc) [inline modules/xfem/src/base/XFEM.C:2189]: I think we still want to rename this method, right?

Comment (bwspenc) [reply modules/xfem/src/base/XFEM.C:2189]: `getElementSideOfInterface`?

Comment (bwspenc) [reply modules/xfem/src/base/XFEM.C:2189]: `getElementSideRelativeToInterface`?

Comment (hugary1995) [reply modules/xfem/src/base/XFEM.C]: I thought we are going to rename this method, along with the data structure and all the logics, in the next PR. Because later on we want to properly manage a vector of cut side IDs.

Anyways, I renamed it just for now.

**Review by bwspenc — DISMISSED** [review]

File: framework/include/materials/MaterialData.h [hunk 9]
```diff
@@ -79,6 +79,22 @@ class MaterialData
...
+   * @param props_old old material property data to copy
+   * @param props_older older material property data to copy, can be empty if the target
+   * element has no older properties
+   * @param side Side number (elemental material properties have this equal to zero)
+   */
+  void copy(const Elem & elem_to,
+            const HashMap<unsigned int, MaterialProperties> & props,
+            const HashMap<unsigned int, MaterialProperties> & props_old,
+            const HashMap<unsigned int, MaterialProperties> & props_older,
+            unsigned int side);
```
Comment (rwcarlsen) [inline framework/include/materials/MaterialData.h:96]: Just to satisfy my curiosity, why do you guys need to manually pass in the hashmaps here?  Why can't you just pass in a material data object like the other copy methods do?

**Review by rwcarlsen — CHANGES_REQUESTED** [review]

File: framework/include/materials/MaterialData.h (hunk 9 above)
Comment (rwcarlsen) [reply framework/include/materials/MaterialData.h:96]: Or actually just pass in nothing (since this is a material data object) - and use the maps from the material data's mat prop storage object?

Comment (hugary1995) [reply framework/include/materials/MaterialData.h:96]: I think at the time we want to copy the material data, the elem_from no longer exists. 

At least that's why I didn't do what you suggested. Maybe you can help us think of a way around, so let me explain to you what we are trying to do:

1. First, there is this `element A`.
2. At some point, we want introduce a cut. So `element A` is cut with XFEM into `partial element B` and `partial element C`. `element A` is deleted.
3. Then, we want to update the location or orientation of the cut, so we first heal the cut. To do that, we delete `partial element B`, and heal `partial element C` to get `element D`.
4. Next, `element D` is re-cut with updated location and orientation of the cut, into `partial element E` and `partial element F`.

Then, obviously partial elements E and F lose the stateful information from partial elements B and C. This PR attempts to fix that. Before healing the cut in step 3, stateful properties from partial elements B and C are stored into a data structure we maintain in XFEM. After re-cutting the element in step 4, previously stored stateful data are copied into partial elements E and F using this new API in MaterialData we added.

With all that being said, it is still possible to avoid extracting and storing material properties from partial elements B and C if we can make them "live longer". In other words, if partial elements B and C are not deleted until we have created partial elements E and F, we can just use the old API to copy around the material properties without introducing this new API. I think @bwspenc can comment on whether this is possible without an overhaul of the current implementation of XFEM.

Comment (rwcarlsen) [reply framework/include/materials/MaterialData.h:96]: Do steps 1 through 4 all happen on the same processor - or could there be a repartitioning or other such elem-moves between any of the steps?  Stateful material properties are stored in a separate data structure - not directly on/with the elements themselves.  So when you delete an element - its stateful property values should still be sitting inside the stateful prop storage data structures.  They are indexed by element pointers.  The problem you cited is that the element object may have been deleted - but it's previous pointer value should still work for indexing those data structures.  The real problem is that our existing APIs for copying/swapping, etc. take element references - when in actuality all they ever do is address the elem to get its underlying pointer/address.  I suspect that if we were to modify the existing APIs, you might not even need to copy around those material data objects - as long as the stateful properties had been synced to the main/sim-global storage loc by the time you want to do this copying.

Comment (bwspenc) [reply framework/include/materials/MaterialData.h:96]: Thanks, @rwcarlsen, we'll do what you are suggesting!  That will considerably simplify things on our end.

Comment (rwcarlsen) [reply framework/include/materials/MaterialData.h:96]: I suppose, however, that there is the risk that element deleting/creating could create a new element pointer with the same memory address as one of the previous elements - an address that was used to index a stateful property.  However, this would only be a problem if in the interrim, stateful properties associated with the new element reusing the old address were stored.  Seems really unlikely to affect you guys.  But this sort of thing is already a problem with our pointer-indexing as it is.  We really should use elem ids instead.  Someday we'll have time to redo all the material prop stuff.

**Comment by hugary1995** [comment]
@rwcarlsen @bwspenc I just switched everything to use the API you suggested. But instead of changing the existing method using element references, I added a new method using pointers.

If we had been doing things correctly this should not break any test.

File: framework/src/materials/MaterialPropertyStorage.C [hunk 10]
```diff
@@ -355,6 +355,26 @@ MaterialPropertyStorage::copy(MaterialData & material_data,
...
 }
 
+void
+MaterialPropertyStorage::copy(MaterialData & material_data,
+                              const Elem * elem_to,
+                              const Elem * elem_from,
+                              unsigned int side,
+                              unsigned int n_qpoints)
+{
+  initProps(material_data, *elem_to, side, n_qpoints);
```
Comment (hugary1995) [inline framework/src/materials/MaterialPropertyStorage.C:365]: Not sure how we should handle potential dereferencing errors here...

Comment (rwcarlsen) [reply framework/src/materials/MaterialPropertyStorage.C:365]: Again - those functions all just need elem pointers - they never actually access the elem instance.

Comment (hugary1995) [reply framework/src/materials/MaterialPropertyStorage.C:365]: Ah you are right. Do you mind if I create another version of initProps that takes in element pointers?

**Review by rwcarlsen — CHANGES_REQUESTED** [review]
All the original copy methods in MaterialData and MaterialPropertyStorage should just call the new functions you created - no need to duplicate all that code (i.e. `copy(const Elem & ...) {copy(&elem_to, &elem_from...);}`).  And since the initProps function is just private/internal - you should just delete the old version and make all calls use the new version.

**Review by rwcarlsen — APPROVED** [review]

**Review by bwspenc — APPROVED** [review]
